### PR TITLE
test(gha): Various Improvements™

### PR DESCRIPTION
**Pull request by joshuarli**
Various improvements.

- py3 backend 4 -> 3 shards; they're a few minutes from the bottleneck job. We reclaim the amount of time it takes to set up one job.
  - reruns enabled now
- use setup-python@v2, they have older minor versions and we can use `.python-version` to dedupe.
  - edit: ha you've got to be kidding me 2.7.16 isn't available, minimum is 2.7.17
  - migration-related jobs will remain pinned to 2.7.17
- deduped test job naming. Example: you'll see `python / 3.6 backend` instead of `python / python3.6 backend`.
  - edit: @billyvg is planning on changing the names soon along with some reorganization
- rb override no longer needed
- removed redundant comments

Also @untitaker how is PYTEST_SENTRY_ALWAYS_REPORT used? org-wide search suggests it isn't. Looks like I didn't ask [here](https://github.com/getsentry/sentry/pull/16938) and just trusted that it did something.

### CODE REVIEW

**Comment by joshuarli** [comment]
Well, don't really want to upgrade 2.7.16 in production. What do people think about changing to 2.7.17 in .python-version?  We've been running 2.7.18 in GHA and 2.7.15 in travis. Forget about travis, but GHA can be made closer to reality.

Or I can ask if upstream will allow 2.7.16.

**Comment by joshuarli** [comment]
Or actually @billyvg, we could remove 2.7 acceptance in GHA given that 3.6 acceptance is required now. It's required by Travis so it is redundant now (you were using it to evaluate GHA initially).

File: requirements-base.txt [hunk 1]
```diff
@@ -110,3 +110,5 @@ phabricator>=0.6.0,<1.0
 # test dependencies, but unable to move to requirements-test until sentry.utils.pytest and similar are moved to tests/
 mock==3.0.5 ; python_version < "3.3"
 selenium==3.141.0
+
+# temporary commit to trigger backend GHA
```
Comment (BYK) [inline requirements-base.txt:114]: Don't forget to remove 🙂

**Review by BYK — APPROVED** [review]

**Comment by joshuarli** [comment]
Oh, check-migration-required runs on 2.7 as well... @wedamija thoughts on just putting that on 3.6?

**Comment by untitaker** [comment]
>Also @untitaker how is PYTEST_SENTRY_ALWAYS_REPORT used?

https://github.com/untitaker/pytest-sentry#always-reporting-test-failures

File: .github/workflows/acceptance.yml [hunk 2]
```diff
@@ -102,156 +99,8 @@ jobs:
...
-        - name: Save snapshots
-          if: always()
-          uses: getsentry/action-visual-snapshot@v2
-          with:
-            save-only: true
-            snapshot-path: .artifacts/visual-snapshots
-
     py3-acceptance:
-      name: 'python3.6 acceptance'
+      name: 'python3.6'
```
Comment (wedamija) [inline .github/workflows/acceptance.yml:103]: Why remove acceptance?

Comment (joshuarli) [reply .github/workflows/acceptance.yml:103]: Because it'll look like `acceptance / python3.6 (0)` as opposed to `acceptance / python3.6 acceptance (0)`.

**Comment by joshuarli** [comment]
Hah, no devservices postgres docker healthcheck. I wanted to do this for redis as well to fix the longstanding issue of `Redis is loading the dataset in memory` if you use redis too soon after devservices up finishes.

This would also be the mechanism to solve the issue with snuba here: https://github.com/getsentry/getsentry/pull/4478#issuecomment-699250885

**Comment by wedamija** [comment]
> Oh, check-migration-required runs on 2.7 as well... @wedamija thoughts on just putting that on 3.6?

I'd almost prefer to run it on both for now, just in case there's a difference between versions. Not sure if we can prevent duplicate messages or whatever.

**Review by billyvg — CHANGES_REQUESTED** [review]
This seems fine, the new acceptance job seems to be missing a step though (I think it's the `Save snapshots`)

**Comment by joshuarli** [comment]
> > Oh, check-migration-required runs on 2.7 as well... @wedamija thoughts on just putting that on 3.6?
> 
> I'd almost prefer to run it on both for now, just in case there's a difference between versions. Not sure if we can prevent duplicate messages or whatever.

Okay, I'm gonna back that out of this PR for now then.

**Comment by joshuarli** [comment]
@billyvg Now that snapshots are saved from py3, they look mostly good, just that there's no differentiation with py2 vs py3-generated.

I looked at all the differences, majority are due to RNG and dictionary reordering. Might have to solve that problem later if we want to diff py2/3 snapshots in the future when py2 travis -> gha.

Only problematic one is `acceptance/organization-releases-with-releases.png`.

**Comment by joshuarli** [comment]
Looks like https://github.com/getsentry/sentry/pull/21128 regressed with py3 testing, temporarily reverting here to run ci.

**Comment by joshuarli** [comment]
Oh, this is going to be a problem since we just flipped the required switch in the repo settings. I'll revert the test name changes, it's not worth the pain.

image

**Comment by billyvg** [comment]
Do we know what's wrong with `acceptance/organization-releases-with-releases.png`?

**Comment by joshuarli** [comment]
> Do we know what's wrong with `acceptance/organization-releases-with-releases.png`?

No idea yet, I haven't looked into it.

**Comment by joshuarli** [comment]
Going to split the py2 acceptance removal + py3 acceptance snapshots out so the other things can be merged earlier.

**Comment by joshuarli** [comment]
Ok, so in the end this PR not a lot of change, but many discoveries were made to follow up on soon. ead2a3c to see backend testing.

**Review by billyvg — APPROVED** [review]

**Comment by joshuarli** [comment]
cool `GitHub Actions has encountered an internal error when running your job.`

really need that feature to only restart select jobs

File: src/sentry/utils/canonical.py [hunk 3]
```diff
@@ -94,7 +94,7 @@ def __init(self, data):
             legacy = settings.PREFER_CANONICAL_LEGACY_KEYS
         norm_func = legacy and get_legacy_name or get_canonical_name
         self._norm_func = norm_func
-        self.data = collections.OrderedDict()
+        self.data = {}
```
Comment (BYK) [inline src/sentry/utils/canonical.py]: Is this intentional?

Comment (joshuarli) [reply src/sentry/utils/canonical.py]: That was just me testing something, it's not part of the final PR diff. You probably didn't refresh your browser.

**Pull request merged**